test/validation/api/event/event.c

```diff
@@ -1,16 +1,557 @@
+
+	for (i = 0; i < EVENT_BURST; i++) {
+		for (j = 0; j < ctx->num_types; j++)
+			mixed[idx++] = ctx->type[j].events[i];
+	}
+
+	num_pkt = odp_event_filter_packet(mixed, packets, remain, idx);
+	CU_ASSERT(num_pkt == EVENT_BURST);
+
+	int pkt_idx = pool_index_of(ODP_POOL_PACKET);
+
+	CU_ASSERT_FATAL(pkt_idx >= 0);
+
+	for (i = 0; i < num_pkt; i++) {
+		CU_ASSERT_FATAL(packets[i] != ODP_PACKET_INVALID);
+
+		event_u64_1 = odp_event_to_u64(odp_packet_to_event(packets[i]));
+		event_u64_2 = odp_event_to_u64(ctx->type[pkt_idx].events[i]);
+		CU_ASSERT(event_u64_1 == event_u64_2);
+	}
+
+	num_rem = idx - num_pkt;
+	idx = 0;
+	for (i = 0; i < EVENT_BURST; i++) {
+		for (j = 0; j < ctx->num_types; j++) {
+			if (j == pkt_idx)
+				continue;
+			expected[idx++] = ctx->type[j].events[i];
+		}
+	}
+	CU_ASSERT(idx == num_rem);
+
+	for (i = 0; i < num_rem; i++) {
+		event_u64_1 = odp_event_to_u64(remain[i]);
+		event_u64_2 = odp_event_to_u64(expected[i]);
+		CU_ASSERT(event_u64_1 == event_u64_2);
+	}
+}
+
+static void event_test_event_frees(void)
+{
+	event_type_ctx_t *ctx = &g_type_ctx;
+	odp_event_t sorted_events_multi[EVENT_BURST];
+	odp_event_t mixed_events_single[MAX_EVENT_TYPES];
+	odp_event_t mixed_events_multi[MAX_EVENT_TYPES];
+	int i, j;
+
+	for (i = 0; i < ctx->num_types; i++) {
+		for (j = 0; j < EVENT_BURST; j++) {
+			sorted_events_multi[j] = event_types_alloc_event(ctx->type[i].pool_type);
+			CU_ASSERT_FATAL(sorted_events_multi[j] != ODP_EVENT_INVALID);
+		}
+
+		odp_event_free_sp(sorted_events_multi, EVENT_BURST);
+	}
+
+	for (i = 0; i < ctx->num_types; i++) {
+		mixed_events_single[i] = event_types_alloc_event(ctx->type[i].pool_type);
+		mixed_events_multi[i] = event_types_alloc_event(ctx->type[i].pool_type);
+		CU_ASSERT_FATAL(mixed_events_single[i] != ODP_EVENT_INVALID);
+		CU_ASSERT_FATAL(mixed_events_multi[i] != ODP_EVENT_INVALID);
+	}
+
+	for (i = 0; i < ctx->num_types; i++)
+		odp_event_free(mixed_events_single[i]);
+
+	odp_event_free_multi(mixed_events_multi, ctx->num_types);
+}
+
+static void event_test_queue_enq_deq(void)
+{
+	event_type_ctx_t *ctx = &g_type_ctx;
+	odp_event_t ev;
+	odp_event_t ev_tbl[MAX_EVENT_TYPES];
+	odp_event_type_t types_seen[MAX_EVENT_TYPES] = {0};
+	odp_event_type_t ev_type;
+	int i, num, total = 0;
+
+	for (i = 0; i < ctx->num_types; i++) {
+		ev = ctx->type[i].events[0];
+		CU_ASSERT(odp_queue_enq(ctx->queue, ev) == 0);
+	}
+
+	while (1) {
+		ev = odp_queue_deq(ctx->queue);
+		if (ev == ODP_EVENT_INVALID)
+			break;
+
+		ev_type = odp_event_type(ev);
+		for (i = 0; i < ctx->num_types; i++)
+			if (ctx->type[i].event_type == ev_type)
+				types_seen[i]++;
+		total++;
+	}
+
+	CU_ASSERT(total == ctx->num_types);
+	for (i = 0; i < ctx->num_types; i++)
+		CU_ASSERT(types_seen[i] == 1);
+
+	for (i = 0; i < ctx->num_types; i++)
+		ev_tbl[i] = ctx->type[i].events[1];
+
+	num = 1;
+	total = 0;
+	while (num > 0 && total < ctx->num_types) {
+		num = odp_queue_enq_multi(ctx->queue, ev_tbl, ctx->num_types);
```

Comment:
If the first call doesn't enqueue all events successfully, the array start pointer and the number of events should be updated accordingly.